pkg/cmd/root/completion.go

```diff
@@ -0,0 +1,236 @@
+	cfg, err := config.LoadFromFile(factory.ConfigFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load configuration: %w", err)
+	}
+
+	names := sets.NewString()
+	for _, garden := range cfg.Gardens {
+		names.Insert(garden.Name)
+	}
+
+	return names.List(), nil
+}
+
+func projectFlagCompletionFunc(cmd *cobra.Command, args []string, toComplete string) ([]string, error) {
+	manager, err := factory.Manager()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create manager: %w", err)
+	}
+
+	// any --garden flag has precedence over the config file
+	var currentTarget target.Target
+
+	if targetProvider.GardenNameFlag != "" {
+		currentTarget = target.NewTarget(targetProvider.GardenNameFlag, "", "", "")
+	} else {
+		currentTarget, err = manager.CurrentTarget()
+		if err != nil {
+			return nil, fmt.Errorf("failed to read current target: %w", err)
+		}
+	}
+
+	gardenClient, err := manager.GardenClient(currentTarget)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Kubernetes client for garden cluster %q: %w", currentTarget.GardenName(), err)
+	}
+
+	projectList := &gardencorev1beta1.ProjectList{}
+	if err := gardenClient.List(factory.Context(), projectList); err != nil {
+		return nil, fmt.Errorf("failed to list projects on garden cluster %q: %w", currentTarget.GardenName(), err)
+	}
+
+	names := sets.NewString()
+	for _, project := range projectList.Items {
+		names.Insert(project.Name)
+	}
+
+	return names.List(), nil
+}
+
+func seedFlagCompletionFunc(cmd *cobra.Command, args []string, toComplete string) ([]string, error) {
+	manager, err := factory.Manager()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create manager: %w", err)
+	}
+
+	// any --garden flag has precedence over the config file
+	var currentTarget target.Target
+
+	if targetProvider.GardenNameFlag != "" {
+		currentTarget = target.NewTarget(targetProvider.GardenNameFlag, "", "", "")
+	} else {
+		currentTarget, err = manager.CurrentTarget()
+		if err != nil {
+			return nil, fmt.Errorf("failed to read current target: %w", err)
+		}
+	}
+
+	gardenClient, err := manager.GardenClient(currentTarget)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Kubernetes client for garden cluster %q: %w", currentTarget.GardenName(), err)
+	}
+
+	seedList := &gardencorev1beta1.SeedList{}
+	if err := gardenClient.List(factory.Context(), seedList); err != nil {
+		return nil, fmt.Errorf("failed to list seeds on garden cluster %q: %w", currentTarget.GardenName(), err)
+	}
+
+	names := sets.NewString()
+	for _, seed := range seedList.Items {
+		names.Insert(seed.Name)
+	}
+
+	return names.List(), nil
+}
+
+func shootFlagCompletionFunc(cmd *cobra.Command, args []string, toComplete string) ([]string, error) {
+	manager, err := factory.Manager()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create manager: %w", err)
+	}
+
+	// for simplicity, always try to read the current target file and ignore any errors
```

Comment:
This comment makes sense when you know how it was previously done. If you do not know the context, you would think that you just spared the error handling as you were too lazy (which was not the case of course) ;-)